packages/pathfinder/src/pathfinder.ts

```diff
@@ -0,0 +1,400 @@
+        asset_code: string
+        asset_issuer: string
+    }
+    >;
+}
+
+interface Cache {
+    cache: any;
+    savePathsToCache(key: string, bucket: number, bestPaths: IPathInfo[]): any;
+    removeFromCache(key: string): any;
+    lookupInCache(key: string): any;
+}
+
+export class PathfinderInitializationError extends Error {}
+
+import * as StellarSdk from 'stellar-sdk'; 
+
+export class Pathfinder {
+
+    /**
+     * This public key should correspond to an account that holds balances of each of
+     * the assets that we support. It will be used as both the destination and source
+     * account for all of the pathfinding algorithms. TODO fill it in
+     */
+    public static REFERENCE_ACCOUNT_PUBLIC_KEY = "GBG47DCYZCY4UU4UN7XSECYT45IRTCUGZN73SVZE5TSCIB3DQDREYUTS";
+    public supportedAssets: stellarAsset[];
+    private server: StellarSdk.Server;
+
+    constructor() {
+        this.server = new StellarSdk.Server('https://horizon.stellar.org');
+    }
+
+    /****************************************
+     * Initialization methods
+     ****************************************/
+
+    /**
+     * Should perform the following actions:
+     *
+     * - loads the list of supported assets
+     */
+    public async init() {
+
+        return Promise.all([
+            this.getSupportedAssets().then(() => this.validateSupportedAssets(this.supportedAssets))
+        ]);
+
+    }
+
+    /**
+     * Loads the list of supported assets from the postgres database
+     */
+    private async getSupportedAssets() {
+
+        let assets = [] as stellarAsset[];
+        const query = {
+            text: 'SELECT id, asset_code, asset_issuer, asset_type from ASSETS'
+        };
+        return pool.query(query)
+            .then((queryResult: any) => queryResult.rows)
+            .then((dbAssets: any) => {
+                assets = dbAssets;
+                this.supportedAssets = dbAssets;
+                return dbAssets;
+            });
+            
+    }
+
+    /**
+     * helper for below
+     * returns a key to be used for lookup in the form of `code-issuer`
+     */
+    private makeKeyPairHelper = (asset_code: string, asset_issuer: string, asset_type: string): string => {
+        const key = (asset_type !== 'native')
+            ? `${asset_code}-${asset_issuer}`
+            : 'XLM';
+        return key;
+    }
+
+
+    /**
+     * Takes in stellarAsset | StellarSdk.Asset 
+     * returns a key to be used for lookup in the form of `code-issuer`
+     */
+    private makeKeyPair = (asset: any): string => {
+        const { asset_code, asset_issuer, asset_type } = asset;
+        return this.makeKeyPairHelper(asset_code, asset_issuer, asset_type);
+    }
+
+    /**
+    * Takes in a StellarSdk.PaymentPathRecord and 
+    * returns a key to be used for lookup in the form of `code-issuer`
+    */
+    private makeKeyPairFromPath = (pathResult: StellarSdk.PaymentPathRecord ): string => {
```

Comment:
Good abstraction of logic. Exactly what I meant when we talked on the phone.

Reply:
thanks!